### DISCUSSION TITLE
interactive notebook

### DISCUSSION THREAD

**Discussion by BenGood** [sharkdp/numbat#362] [Q&A]
Is there any way to use numbat like soulver or [notecalc](https://bbodi.github.io/notecalc3/notecalc#eJx9Vttu2zgQfedXDNBLLlVtyY7TeLd9SBp0W6BJiiTt7iLIApRF24QlUiAlO-7X7xlSdpwGaALbEi9zOXN4hpe2URNZTkh7kjSXplgTv7elbKyjxq21mVFjKXfhYa5IFktpGjlTnuyUbmxbLpUTWMKTK5X3hPjXtrBiqPWKdEOSbU9slWsjG20N75O7buAWAwax1LJIqNIP7GyzADt8Ikxb5cr5hGytHO_CY2t0E6KolPStU5UyDa10M-cBAxvTtkyoUH7idN3opaJGPWCHdLaFSwRcJaJ2dqmL4NAanqcpQlrb1j2JoEfXyrcl-zMhVadnc-TmgEjb2AqrsLxci7YuZKMKWs2VCf5oAlwBF4C55VeGencHebnEem2i09JiFHjblVcuYVTmHBx2eU6vAxpzgH0H65mTeQCyspznNsBZ2zTKEe3z0MwpuAPKAUEqtVGUq2alONQ5XKhCY0LIiA65mPJBBJVHKstFRRCY0j-jo7ipHxczIhJx3dgKMw-yqkukTp-UKmnqVNgbAQn4h5DrUq67omCneEHflJuoSDKRpSm9oSx9JQZ4OqRRfHhLAwzx4svIjIBEYc1eAx4vox9YbOinctbDQLoQhD_knaUX4qaWE7XBAvB3_MIcpWn4CHGksoEY9rIj9XYQXJ2BwW4dYv6sHsTT162J9OHTJ5HmWYw8zdM0y0SOg-DnetpwMXkszej9ezoS2TH9c3VNg9FIXF7d7uP3gE4vz4MV2BG3QDiXAB08f6xJqHmuOigLytdUY8az9a9q2nCa14EAk9Z5vKzmulRibstA9cKuDJ3CzEKte8Qu9tKHvZDJXprvPeV4R6FGlaVnjjJVYW4yD3GJnaBkXSvpfCzitZoqp8yE_THT_JarTjGJdk8Xxz8p9WTRqUzFPnYM19KoMhGhel6VatJsF26p-f1b__zq70tOycd0aZPu6ddbxJQNkHdZSgfgCjBpyL8eAqDBBJB_SLWyoCumXt8dndwL8SUGIXM-U5xDQsPjqJWuyw6WlrJsFchtq3iihic9Eh8jx6VZd-fXbCaTALNmpSpLriFMcUZRA2LSj8gF470tdGqJo_p4xB-DCAg_OUQ_pNMyx-kTZ9IsmD6nlUKukj7QKDD8ZX-NcoGjo94Ax-qjbnTOS5_NH78S53ra-WJL26Vv6Rfj4vXdaHzP6PJej0Be0CnkDvoPaLnYwY-42OTQ8VM2UUsATUd08HmpITcRt6ATzHC6PT3rhVOx7BIkI6E1GzSfamvnVjEm2MNqjY9nccpbDWYBcm80eMv-5fZcyceQFVqS6s16oldr0YPpodjvIoHkRqLtJMLs-H0u_iCgctNW2xTAzZeh7-QKzXRwmA3jK9JYqMYL5EK-hkGmxAfybSUEvkiWKwm2M8997AsY_NXnkzYmPitZsEohA-Qcv7mNRhZDNVkvewjwBXQByczDepw6rGHr3QqRhgCwwNhVHGHSfeeWHKrjrEQTsXZBR6NKm7ZhenZ9BamoLTZLBS0BhivcEw4pGyyq_lzsB9-Mpdl0ka4-tuI277eFMutVqCsKGdZGGQyp3j5j0cq6BcQGd57ubgIPELGOLJK8Dp4qaI6uoUcBNLFpkU8TiZEm4FEod0cteCj8TlNV8cJiirBJbAfQMZOoogkDgVblPOSFTeiZsW5zg9gWr7t-RUE8EOdQ8BIQ87ik7K8zmjIRgxr2ucPlfc-YIGLBosmJrcO6X2chi5xIeO17MaLqv0GcXfDTnMt6IRunH0LaPyBVuCLcjZLjNHmXnPBZv8uSlP__TJNB95smw-43TY7uiY187ErHHt9Abfv9kNGmbbLxCI22_LqDBGLZ1P0P4M7qEtj2uZ1FZbip-IR0lsJpqS1YkOsSWoN6faDx-HB8cjh-dzg-PhyPAjW4d04C5VfamA3ApeVut2ZmZP1nlsQXvrvGY4mYPIg4hFCKvI03ZaZo1aI1dktWuB8XYSE9j4r7TJaO7tloPLF8peqOxU4cw96xGr-k_dur8yt-oQtXHHAKpqPwPts5vkdp-eHd_QHcVwi2Anp8JJTxOl6JEGcIhD104sJBb_rOSjpuTWix0574H6YbB4c) ?

I'd like to be able to change things and see the output of every line update.

**Comment by triallax** [comment]
There's a WIP PR for Jupyter integration: https://github.com/sharkdp/numbat/pull/147

Maybe you'll find it useful (I don't know if Jupyter shows you the output of every line though...)

**Comment by sharkdp** [comment]
> Is there any way to use numbat like soulver or notecalc ?

No. But this could be a fun thing to try! Numbat is certainly fast enough to support this use case (worst case: reevaluate the entire program on every keypress). I'm currently playing with a [mobile interface](https://github.com/sharkdp/numbat/issues/218#issuecomment-1957084709) that also updates a result on every keypress. But the layout is not soulver-like.

**Reply by BenGood** [reply]
very cool. 
I had the same thought about recalculating everything all the time and started playing with this powershell script. It runs every line in the nbt script through numbat -e and then adds a line under the expression and inserts the result of that expression as a comment. Then it overwrites the whole script. Since the results are commented, the script is still valid. 

If I set that to run continually I can open script.nbt in vscode and every time I save it, the calculations are made and vscode refreshes to show the updates. 

```powershell
$updatednbt ="" 
get-content script.nbt|foreach-object{ 
    if ($_ -match '^\s*#!') {
        ; # result of previous run to be discarded
    } 
    elseif ($_ -match '(^\s*#|^\s*$)'){
        
        $updatednbt= $updatednbt+$_ + "`n"; # some other comment to keep
    }
    else{ # nbt expression to calculate
        $updatednbt= $updatednbt+$_ + "`n";
        $result = invoke-expression "numbat -e '$_'"
        $updatednbt= $updatednbt + "`t#!" 
        $updatednbt= $updatednbt + $result.Replace("`n","").Replace("`r","")+ "`n";
    }
}
Set-Content .\script.nbt -Value $updatednbt

```

**Reply by BenGood** [reply]
never mind. that approach could not work with let statements. 

Everyone has a different idea for where to go with an awesome project like this. I'll be interested to see how it develops.

**Comment by michelk** [comment]
A Jupyter-Kernel would be great for interactive formula-collections.

**Reply by sharkdp** [reply]
see also: https://github.com/sharkdp/numbat/pull/147
